### PR TITLE
Fix missing generics arity error

### LINKED ISSUE #6507: Missing static error for class descending from `T::Generic` given too many type arguments
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20NotABox%0A%20%20extend%20T%3A%3AGeneric%0Aend%0A%0Asig%20%7Bparams%28x%3A%20NotABox%5BInteger%5D%29.void%7D%0Adef%20example1%28x%29%0A%20%20T.reveal_type%28x%29%0Aend%0A%0Aclass%20Box%0A%20%20extend%20T%3A%3AGeneric%0A%0A%20%20Elem%20%3D%20type_member%0Aend%0A%0Asig%20%7Bparams%28x%3A%20Box%5BInteger%2C%20String%5D%29.void%7D%0Adef%20example2%28x%29%0A%20%20T.reveal_type%28x%29%0Aend)

```ruby
# typed: true
extend T::Sig

class NotABox
  extend T::Generic
end

sig {params(x: NotABox[Integer]).void}
def example1(x)
  T.reveal_type(x)
end

class Box
  extend T::Generic

  Elem = type_member
end

sig {params(x: Box[Integer, String]).void}
def example2(x)
  T.reveal_type(x)
end
```

#### Observed output

```
editor.rb:19: Wrong number of type parameters for Box. Expected: 1, got: 2 https://srb.help/7010
    19 |sig {params(x: Box[Integer, String]).void}
                           ^^^^^^^^^^^^^^^

editor.rb:10: Revealed type: T.class_of(NotABox)[Runtime object representing type: Integer] (unresolved) https://srb.help/7014
    10 |  T.reveal_type(x)
          ^^^^^^^^^^^^^^^^
  Got T.class_of(NotABox)[Runtime object representing type: Integer] (unresolved) originating from:
    editor.rb:9:
     9 |def example1(x)
                     ^

editor.rb:21: Revealed type: Box[Integer] https://srb.help/7014
    21 |  T.reveal_type(x)
          ^^^^^^^^^^^^^^^^
  Got Box[Integer] originating from:
    editor.rb:20:
    20 |def example2(x)
                     ^

editor.rb:19: Wrong number of type parameters for Box. Expected: 1, got: 2 https://srb.help/7010
    19 |sig {params(x: Box[Integer, String]).void}
                           ^^^^^^^^^^^^^^^
Errors: 4
```

`Box` expects one type argument. When given two, an error is reported. (This is correct.)

#### Expected behavior

`NotABox` is not actually a generic class (it has no type members). When given a type argument, no errors is raised (this is a bug).
- - -

### PR DESCRIPTION
**Pull request by jez**
The idea is that we're not supposed to short circuit this function, because it
reports an arity mismatch error we care about.

While I was there, I decided to make the error a little more informative.

### Motivation

Fixes #6507

### Test plan

See included automated tests.

### CODE REVIEW

File: test/testdata/infer/generics/arity.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# typed: true
+extend T::Sig
```
Comment (neilparikh) [inline test/testdata/infer/generics/arity.rb:2]: Can you add a test for the cases where `NotABox` and `Box` are passed the current amount of generic args?

Comment (jez) [reply test/testdata/infer/generics/arity.rb:2]: :portal-from: DM: these tests are already covered by the rest of Sorbet's test suite and we're going to opt to not add more redundant tests here.

File: test/testdata/infer/generics/arity.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# typed: true
```
Comment (neilparikh) [inline test/testdata/infer/generics/arity.rb:1]: Looks like this test duplicates`test/testdata/infer/generics/arity_mismatch.rb`, can you merge the two?

**Review by neilparikh — CHANGES_REQUESTED** [review]

**Comment by neilparikh** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Nd4rWKuc5FfFqW
→ https://go/builds/bui_Nd4rNOI8R4s7BV

**Review by neilparikh — APPROVED** [review]

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Nd9vWSChv0eyBR
→ https://go/builds/bui_Nd9wGMDf3Ka2UJ

**Pull request merged**